File: data_updates/R/iati_rhfp.R
####Setup###/#
list.of.packages <- c("reshape2","data.table","openxlsx","plyr","gdata","varhandle","rsdmx","rstudioapi","RCurl","git2rdata")
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages)
lapply(list.of.packages, require, character.only=T)

wd <- dirname(getActiveDocumentContext()$path)
setwd(wd)

#bring in data
options(timeout = 1000)
temp <- tempfile()
download.file("https://ddw.devinit.org/api/export/1231",temp,method = "libcurl")
dat <- fread(temp)
names(dat) = c("year","Reporting Organisation Narrative","aid_type_di_name","Sector Code - Calculated", "DAC3 Sector Code - Calculated","Reporting Organisation Type Code","Reporting Organisation Type Name - Calculated","Recipient Code","Recipient Name","Flow Type Code - Calculated","Transaction Receiver Organisation Type","Transaction Type","x_transaction_value_usd_m_Sum")
dat <- dat[!(dat$`Sector Code - Calculated` == 13010 |dat$`Sector Code - Calculated` == 13081),]

#Add necessary Columns
dat <- mutate(dat, purpose_name = ifelse(`Sector Code - Calculated` == 13030, "Family planning", ifelse(`Sector Code - Calculated` == 13020, "Reproductive health care", "NA")))
dat <- mutate(dat, `Channel of delivery` = ifelse(`Transaction Receiver Organisation Type` == 21, "NGOs and Civil Society", ifelse(`Transaction Receiver Organisation Type` == 40, "Multilateral", ifelse(`Transaction Receiver Organisation Type` == 10, "Public Sector", ifelse(`Transaction Receiver Organisation Type` == 70, "Private Sector", ifelse(`Transaction Receiver Organisation Type` == 80, "Universities and Research Institutes", ifelse(`Transaction Receiver Organisation Type` == 22,"NGOs and Civil Society", ifelse(`Transaction Receiver Organisation Type` == 90, "Other",ifelse(`Transaction Receiver Organisation Type` == 73, "Private Sector", ifelse(`Transaction Receiver Organisation Type` ==23, "NGOs and Cvil Society", ifelse(`Transaction Receiver Organisation Type` == 24, "NGOs and Civil Society","Not Reported")))))))))))

#Get transactions in correct units
dat$x_transaction_value_usd_m_Sum <- dat$x_transaction_value_usd_m_Sum/1000000

repo <- repository("C:/git/di-website-data")
pull(repo)
write_vc(dat, file = "2022/iati_rhfp.csv", root = repo, stage = TRUE)
commit(repo, "Update of the RHFP data for Gates")
push(repo)

File: data_updates/Python/iati_rhfp.py
import os
import requests
import glob
from requests.auth import HTTPBasicAuth
import pandas as pd
from utils.data_updates import push_folder_to_github

try:
    os.mkdir("iati_csv")
except:
    print("Folder already exists!")

donors_selected = pd.read_csv("https://ddw.devinit.org/api/export/1447")

donors_selected.columns = ["reporting_org_ref","registry_id","org_type","org_name","abbreviation","usability_score","tracker_commit","tracker_spend","latest_txn_year","hq_country","CRS_codes","IATI_organisation_type","IATI_organisation_name","country"]

donors_selected = donors_selected[donors_selected['tracker_spend'] == "Yes"]

donors_selected = donors_selected[['reporting_org_ref','country']]


# Dataset 1

data = pd.read_csv("https://ddw.devinit.org/api/export/1231")

data = pd.DataFrame(data)

data.columns = ["Reporting Organisation Reference","Reporting Organisation Narrative","year","Sector Code - Calculated"
, "DAC3 Sector Code - Calculated","Reporting Organisation Type Code","Reporting Organisation Type Name - Calculated"
,"Recipient Code","Recipient Name","Flow Type Code - Calculated","aid_type_di_name","x_transaction_value_usd_m_Sum"]

data = data[data['Reporting Organisation Reference'].isin(donors_selected['reporting_org_ref'])]

data = data[(data['Sector Code - Calculated'] != 13010) & (data['Sector Code - Calculated'] != 13081)]

data = pd.merge(data, donors_selected, left_on='Reporting Organisation Reference',right_on='reporting_org_ref').drop('reporting_org_ref', axis=1)

data = data.groupby(["year","country","Sector Code - Calculated"
, "DAC3 Sector Code - Calculated","Reporting Organisation Type Code","Reporting Organisation Type Name - Calculated"
,"Recipient Code","Recipient Name","Flow Type Code - Calculated","aid_type_di_name"],dropna=False).sum().reset_index()

# Sector code mapping

data['purpose_name'] = ""

data.loc[data['Sector Code - Calculated']==13020,'purpose_name'] = "Reproductive health care"
data.loc[data['Sector Code - Calculated']==13030,'purpose_name'] = "Family planning"

# Only ODA

data = data[data['Flow Type Code - Calculated'] ==10]

# Put it in millions

data["x_transaction_value_usd_m_Sum"] = data["x_transaction_value_usd_m_Sum"]/1000000

# Rename away from country

data["Reporting Organisation Narrative"] = data["country"]

data = data[data['country'] != "Germany"]

data.to_csv("iati_csv/IATI-RHFP-data-v1.csv", encoding='utf-8', index=False)

# Dataset 2

data = pd.read_csv("https://ddw.devinit.org/api/export/1271")

data = data[data['Reporting Organsation Reference'].isin(donors_selected['reporting_org_ref'])]

data = pd.DataFrame(data)

data.columns = ["Code type","Year", "recipient_name","Reporting Organisation Narrative","Reporting Organisation Reference","Flow Type Code - Calculated","disbursed"]

data = data[(data['Code type'] != 13010) & (data['Code type'] != 13081)]

# Sector code mapping

data.loc[data['Code type']==13020,'Code type'] = "Reproductive health care"
data.loc[data['Code type']==13030,'Code type'] = "Family planning"

# Only ODA

data = data[data['Flow Type Code - Calculated'] ==10]

data = pd.merge(data, donors_selected, left_on='Reporting Organisation Reference',right_on='reporting_org_ref').drop('reporting_org_ref', axis=1)

# Summing both

total_data = data.groupby(["Year","recipient_name","country"]).agg({"disbursed":"sum"}).reset_index()
total_data['Code type'] = "Reproductive health care and family planning"

# Groupby country

data = data.groupby(["Year","recipient_name","country","Code type"]).agg({"disbursed":"sum"}).reset_index()

data = data.append(pd.DataFrame(data = total_data),ignore_index=True)

# Put it in millions

data["disbursed"] = data["disbursed"]/1000000

data = data.pivot(index=['Code type','country','recipient_name'], columns='Year', values='disbursed').reset_index()

data["Reporting Organisation Narrative"] = data["country"]

data = data[data['country'] != "Germany"]

data.to_csv("iati_csv/IATI-RHFP-data-v2.csv", encoding='utf-8', index=False)

# Dataset 4

data = pd.read_csv("https://ddw.devinit.org/api/export/1309")

data = pd.DataFrame(data)

data.columns = ["Donor Name","Aid Type Di Name", "Year","Purpose Code","Reporting Organisation Reference", "Usd Disbursement Deflated Sum"]

data = data[data["Reporting Organisation Reference"].isin(donors_selected['reporting_org_ref'])]

data = pd.merge(data, donors_selected, left_on='Reporting Organisation Reference',right_on='reporting_org_ref').drop('reporting_org_ref', axis=1)

data = data[(data['Purpose Code'] != 13010) & (data['Purpose Code'] != 13081)]

# Sector code mapping

data.loc[data['Purpose Code']==13020,'Purpose Name'] = "Reproductive health care"
data.loc[data['Purpose Code']==13030,'Purpose Name'] = "Family planning"

# Put it in millions

data["Usd Disbursement Deflated Sum"] = data["Usd Disbursement Deflated Sum"]/1000000

data = data.groupby(["Year","Aid Type Di Name","country","Purpose Name","Purpose Code"]).agg({"Usd Disbursement Deflated Sum":"sum"}).reset_index()

data["Donor Name"] = data["country"]

data = data[data['country'] != "Germany"]

data.to_csv("iati_csv/iati_rhfp4.csv", encoding='utf-8', index=False)

# Sample call to the function below
push_folder_to_github('devinit/di-website-data', 'main', 'iati_csv', '2022', 'Committing from API', '*.csv')
